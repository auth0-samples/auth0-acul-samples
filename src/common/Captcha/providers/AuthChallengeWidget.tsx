import React, { useState } from "react";
import Turnstile from "react-turnstile";
import type { CaptchaWidgetProps, CaptchaResponse } from "../index";
import { cn } from "@/utils/helpers/cn";

const AuthChallengeWidget: React.FC<CaptchaWidgetProps> = ({
  config,
  onCaptchaResponse,
  className = "",
}) => {
  const [token, setToken] = useState("");

  if (config.provider !== "auth0_v2") {
    return null;
  }

  const siteKey = config.siteKey;

  if (!siteKey) {
    return null;
  }

  const handleVerify = (token: string) => {
    setToken(token);
    const response: CaptchaResponse = {
      provider: "auth0_v2",
      token: token,
    };
    onCaptchaResponse(response);
  };

  const handleError = () => {
    setToken("");
    onCaptchaResponse(null);
  };

  const handleExpire = () => {
    setToken("");
    onCaptchaResponse(null);
  };

  return (
    <div className={cn("space-y-2", className)}>
      <div className="w-full flex justify-center">
        <Turnstile
          sitekey={siteKey}
          onVerify={handleVerify}
          onError={handleError}
          onExpire={handleExpire}
          theme="light"
          size="normal"
        />
        <input type="hidden" name="captcha" value={token} />
      </div>
    </div>
  );
};

export default AuthChallengeWidget;
