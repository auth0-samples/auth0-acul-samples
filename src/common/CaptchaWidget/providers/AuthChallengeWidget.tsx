import React, { useState } from "react";
import FormField from "@/common/FormField";
import Button from "@/common/Button";
import type { InputProps } from "@/common/Input";
import type { LabelProps } from "@/common/Label";
import { cn } from "@/utils/helpers/cn";
import type { CaptchaWidgetProps, CaptchaResponse } from "../types";

const AuthChallengeWidget: React.FC<CaptchaWidgetProps> = ({
  config,
  onCaptchaResponse,
  onError,
  className = "",
  label,
  error,
}) => {
  const [code, setCode] = useState("");
  const [challengeId, setChallengeId] = useState<string | null>(null);
  const [isChallengeSent, setIsChallengeSent] = useState(false);
  const [isLoading, setIsLoading] = useState(false);

  if (config.provider !== "auth-challenge") {
    return null;
  }

  const challengeType = config.challengeType;
  const endpoint = config.endpoint || "/api/auth/challenge";

  const displayLabel = label || `Enter ${challengeType.toUpperCase()} code*`;
  const challengeButtonText = `Send ${challengeType.toUpperCase()} Code`;

  const sendChallenge = async () => {
    setIsLoading(true);
    try {
      const response = await fetch(endpoint, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          type: challengeType,
          action: "send",
        }),
      });

      if (response.ok) {
        const data = await response.json();
        setChallengeId(data.challengeId);
        setIsChallengeSent(true);
      } else {
        onError?.(`Failed to send ${challengeType} challenge`);
      }
    } catch (error) {
      console.error("Challenge send error:", error);
      onError?.(`Error sending ${challengeType} challenge`);
    } finally {
      setIsLoading(false);
    }
  };

  const handleCodeChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const value = event.target.value;
    setCode(value);

    if (challengeId) {
      const response: CaptchaResponse = {
        provider: "auth-challenge",
        answer: value,
        challengeId: challengeId,
      };
      onCaptchaResponse(response);
    }
  };

  const formFieldLabelProps: LabelProps = {
    children: displayLabel,
    htmlFor: "auth-challenge-input",
  };

  const formFieldInputProps: InputProps = {
    id: "auth-challenge-input",
    name: "auth-challenge",
    type: "text",
    value: code,
    onChange: handleCodeChange,
    placeholder: "Enter verification code...",
    autoComplete: "one-time-code",
    disabled: !isChallengeSent,
  };

  return (
    <div className={cn("auth-challenge-container space-y-3", className)}>
      {/* Challenge initiation */}
      {!isChallengeSent && (
        <div className="space-y-3">
          <p className="text-body-text text-sm">
            We'll send a verification code to verify it's you.
          </p>
          <Button
            type="button"
            variant="secondary"
            onClick={sendChallenge}
            isLoading={isLoading}
            loadingText="Sending..."
            fullWidth
          >
            {challengeButtonText}
          </Button>
        </div>
      )}

      {/* Code input (appears after challenge is sent) */}
      {isChallengeSent && (
        <div className="space-y-3">
          <p className="text-body-text text-sm text-success">
            Verification code sent! Please check your{" "}
            {challengeType === "sms" ? "messages" : challengeType}.
          </p>
          <FormField
            labelProps={formFieldLabelProps}
            inputProps={formFieldInputProps}
            error={error}
          />
        </div>
      )}
    </div>
  );
};

export default AuthChallengeWidget;
