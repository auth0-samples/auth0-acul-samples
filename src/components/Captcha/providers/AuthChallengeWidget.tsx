import React, { useEffect, useRef, useState } from "react";
import Turnstile from "react-turnstile";

import { cn } from "@/lib/utils";

import type { CaptchaResponse, CaptchaWidgetProps } from "../index";

const MAX_RETRY_ATTEMPTS = 3; // Define a maximum number of retries

const AuthChallengeWidget: React.FC<CaptchaWidgetProps> = ({
  config,
  onCaptchaResponse,
  className = "",
  theme,
  error, // what to do with backend error
}) => {
  const [errorMessage, setErrorMessage] = useState<string | undefined>(error); // Manage internal as well as error from backend
  const retryCount = useRef(0); // Use a ref to persist retry count across renders

  console.log(errorMessage);

  // Update internal error if external error prop changes
  useEffect(() => {
    setErrorMessage(error);
  }, [error]);

  if (config.provider !== "auth0_v2") {
    return null;
  }

  const siteKey = config.siteKey;

  if (!siteKey) {
    return null;
  }

  const handleVerify = (token: string) => {
    retryCount.current = 0; // Reset retry count on successful verification
    setErrorMessage(undefined); // Clear any previous error

    const response: CaptchaResponse = {
      provider: "auth0_v2",
      token: token,
    };
    onCaptchaResponse(response);
  };

  const handleError = (error: string) => {
    onCaptchaResponse(null);

    if (retryCount.current < MAX_RETRY_ATTEMPTS) {
      retryCount.current += 1;
      return;
    } else {
      console.log("error from client", error);

      setErrorMessage(`Verification failed after multiple attempts: ${error}`);
    }
  };

  const handleExpire = () => {
    onCaptchaResponse(null);
    retryCount.current = 0; // Reset retry count on expiration
  };

  const auth0CaptchaWidgetStyle = {
    transform: "scale(1.06)",
    transformOrigin: "0 0",
  };

  return (
    <div className={cn("space-y-2", className)}>
      <div className="w-full justify-between" style={auth0CaptchaWidgetStyle}>
        <Turnstile
          sitekey={siteKey}
          onVerify={handleVerify}
          onError={handleError}
          onExpire={handleExpire}
          theme={theme}
          size="normal"
        />
      </div>
      {/* {errorMessage && (
        <p className="text-sm font-medium text-destructive">{errorMessage}</p>
      )} */}
    </div>
  );
};

export default AuthChallengeWidget;
