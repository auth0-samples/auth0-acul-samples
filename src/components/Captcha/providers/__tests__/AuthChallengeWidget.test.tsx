import { FormProvider, useForm } from "react-hook-form";

import { render, screen } from "@testing-library/react";

// Mock Turnstile so it renders a testable element
jest.mock("react-turnstile", () => ({
  __esModule: true,
  default: (props: any) => <div role="presentation" {...props} />,
}));

import AuthChallengeWidget from "../AuthChallengeWidget";

const baseProps = {
  config: {
    provider: "auth0_v2",
    siteKey: "test-site-key",
  },
  onCaptchaResponse: jest.fn(),
  className: "",
  theme: "light" as const,
  control: undefined,
  name: "captcha",
};

function Wrapper({
  children,
  defaultValues = {},
}: {
  children: any;
  defaultValues?: any;
}) {
  const methods = useForm({ defaultValues });
  return (
    <FormProvider {...methods}>
      <form>{children({ control: methods.control })}</form>
    </FormProvider>
  );
}

describe("AuthChallengeWidget", () => {
  it("renders the Turnstile widget", () => {
    render(
      <Wrapper>
        {({ control }: { control: any }) => (
          <AuthChallengeWidget {...baseProps} control={control} />
        )}
      </Wrapper>
    );
    expect(screen.getByRole("presentation")).toBeInTheDocument();
  });

  it("calls onCaptchaResponse on verify", () => {
    // Simulate Turnstile onVerify
    const onCaptchaResponse = jest.fn();
    render(
      <Wrapper>
        {({ control }: { control: any }) => (
          <AuthChallengeWidget
            {...baseProps}
            control={control}
            onCaptchaResponse={onCaptchaResponse}
          />
        )}
      </Wrapper>
    );
    expect(typeof onCaptchaResponse).toBe("function");
  });

  it("matches snapshot (default)", () => {
    const { container } = render(
      <Wrapper>
        {({ control }: { control: any }) => (
          <AuthChallengeWidget {...baseProps} control={control} />
        )}
      </Wrapper>
    );
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot (with error)", () => {
    const { container } = render(
      <Wrapper>
        {({ control }: { control: any }) => (
          <AuthChallengeWidget
            {...baseProps}
            control={control}
            error="Test error"
          />
        )}
      </Wrapper>
    );
    expect(container).toMatchSnapshot();
  });
});
