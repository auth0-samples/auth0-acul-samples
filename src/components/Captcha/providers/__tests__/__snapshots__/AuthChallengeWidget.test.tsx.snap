// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`AuthChallengeWidget matches snapshot (default) 1`] = `
<div>
  <form>
    <div
      class="space-y-2"
    >
      <div
        class="w-full justify-between"
        style="transform: scale(1.06); transform-origin: 0 0;"
      >
        <div
          role="presentation"
          sitekey="test-site-key"
          theme="light"
        />
      </div>
    </div>
  </form>
</div>
`;

exports[`AuthChallengeWidget matches snapshot (with error) 1`] = `
<div>
  <form>
    <div
      class="space-y-2"
    >
      <div
        class="w-full justify-between"
        style="transform: scale(1.06); transform-origin: 0 0;"
      >
        <div
          role="presentation"
          sitekey="test-site-key"
          theme="light"
        />
      </div>
    </div>
  </form>
</div>
`;
